src/instreamTracking.js

```diff
@@ -0,0 +1,106 @@
+import { config } from './config.js';
+import { auctionManager } from './auctionManager.js';
+import { VIDEO } from './mediaTypes.js';
+import { OUTSTREAM } from './video.js';
+import * as events from './events.js';
+import * as utils from './utils.js';
+import { TARGETING_KEYS, EVENTS, BID_STATUS } from '../src/constants.json';
+
+const { CACHE_ID, UUID } = TARGETING_KEYS;
+const { BID_WON } = EVENTS;
+const { RENDERED } = BID_STATUS;
+
+const INSTREAM_TRACKING_DEFAULT_CONFIG = {
+  enabled: false,
+  maxWindow: 1000 * 60, // the time in ms after which polling for instream delivery stops
+  pollingFreq: 500 // the frequency of polling
+};
+
+// Set instreamTracking default values
+config.setDefaults({
+  'instreamTracking': utils.deepClone(INSTREAM_TRACKING_DEFAULT_CONFIG)
+});
+
+const whitelistedResources = /video|fetch|xmlhttprequest|other/;
+
+/**
+ * Here the idea is
+ * find all network entries via performance.getEntriesByType()
+ * filter it by video cache key in the url
+ * and exclude the ad server urls so that we dont match twice
+ * eg:
+ * dfp ads call: https://securepubads.g.doubleclick.net/gampad/ads?...hb_cache_id%3D55e85cd3-6ea4-4469-b890-84241816b131%26...
+ * prebid cache url: https://prebid.adnxs.com/pbc/v1/cache?uuid=55e85cd3-6ea4-4469-b890-84241816b131
+ *
+ * if the entry exists, emit the BID_WON
+ *
+ * Note: this is a workaround till a better approach is engineered.
+ *
+ * @param {Array<AdUnit>} adUnits
+ * @param {Array<Bid>} bidsReceived
+ *
+ * @return {boolean} returns TRUE if tracking started
+ */
+export function trackInstreamDeliveredImpressions(adUnits, bidsReceived) {
+  const instreamTrackingConfig = config.getConfig('instreamTracking') || {};
+  // check if instreamTracking is enabled and performance api is available
+  if (!instreamTrackingConfig.enabled || !window.performance || !window.performance.getEntriesByType) {
+    return false;
+  }
+
+  // filter for video bids
+  const instreamBids = bidsReceived.filter(bid => bid.mediaType === VIDEO && bid.context !== OUTSTREAM && bid.videoCacheKey);
+  if (!instreamBids.length) {
+    return false;
+  }
+
+  // find unique instream ad units
+  const instreamAdUnitMap = {}
+  adUnits.forEach(adUnit => {
+    if (!instreamAdUnitMap[adUnit.code] && utils.deepAccess(adUnit, 'mediaTypes.video.context') !== OUTSTREAM) {
+      instreamAdUnitMap[adUnit.code] = true;
+    }
+  });
+  const instreamAdUnitsCount = Object.keys(instreamAdUnitMap).length;
+
+  const start = Date.now();
+  const { maxWindow, pollingFreq } = instreamTrackingConfig;
+
+  let instreamWinningBidsCount = 0;
+  let lastRead = 0; // offset for performance.getEntriesByType
+
+  function poll() {
+    // get network entries using the last read offset
+    const entries = window.performance.getEntriesByType('resource').splice(lastRead);
```

Comment:
Would there ever be a time when this code is executed that the `window.performance.getEntriesByType('resource')` would not return any values?  If so - we should add the proper conditional handlers to the logic here in case there are empty values (either here or later on in the function).

Reply:
It would will always give some entries, as this returns network entries. So something or the other would be loading on the page.  
Eg: publisher css, script, prebid.js script, etc